src/main/java/seedu/address/logic/parser/TaggingCommandParserUtil.java

```diff
@@ -0,0 +1,52 @@
+package seedu.address.logic.parser;
+
+import static seedu.address.logic.Messages.MESSAGE_INVALID_COMMAND_FORMAT;
+import static seedu.address.logic.parser.CliSyntax.PREFIX_TAG;
+
+import java.util.HashSet;
+import java.util.Set;
+import java.util.stream.Stream;
+
+import javafx.util.Pair;
+import seedu.address.commons.core.index.Index;
+import seedu.address.logic.parser.exceptions.ParseException;
+import seedu.address.model.tag.Tag;
+
+/**
+ * A utility class that provides methods for parsing input arguments related to
+ * tagging commands, such as extracting the index and tags from user input.
+ */
+public class TaggingCommandParserUtil {
+
+    /**
+     * Parses the index and tags from the argument map.
+     *
+     * @param argMultimap the argument multimap containing the preamble and tags
+     * @return a pair consisting of the parsed index and tag list
+     * @throws ParseException if the index or tags are invalid
+     */
+    public static Pair<Index, Set<Tag>> parseIndexAndTags(ArgumentMultimap argMultimap, String messageUsage)
+            throws ParseException {
+        Index index;
+        if (!arePrefixesPresent(argMultimap, PREFIX_TAG)) {
+            throw new ParseException(String.format(MESSAGE_INVALID_COMMAND_FORMAT, messageUsage));
+        }
+
+        try {
+            index = ParserUtil.parseIndex(argMultimap.getPreamble());
+        } catch (ParseException pe) {
+            throw new ParseException(String.format(MESSAGE_INVALID_COMMAND_FORMAT, messageUsage), pe);
```

Comment:
I like how you did the abstraction with the message usage being passed in by whatever tag/untag command calls this method